Added notification after paste content from external sources (Word, Excel etc.) to the examples

KarolDawidziuk opened the pull request:
I've added notification when a user tries to paste content from external sources (Word, Excel, GDocs, LibreOffice) into the examples. 
The code detects the source from which the content is being pasted and, based on that, displays appropriate links to the examples. 

For some reasons full detection doesn't working in Safari and IE.
Safari detects pasted content from Word and Google Docs.
IE can only detect pasted content from Word.

jacekbogdanski commented on docs/sdk/examples/assets/pastefromoffice/pastefromoffice.js:
```diff
@@ -0,0 +1,164 @@
+        isAlreadyShow: false
+    }
+];
+var instances = getInstances();
+// A state variable that indicates if the clipboard notification has been seen.
+// We use the variable for displaying the notification only once per demo.
+let hasNotificationBeenSeen = false;
```
Mixed `var` and `let` keywords. Let's stick to `var` as this JS is compiled to support older browsers.

jacekbogdanski commented on docs/sdk/examples/assets/pastefromoffice/pastefromoffice.js:
```diff
@@ -0,0 +1,164 @@
+// We use the variable for displaying the notification only once per demo.
+let hasNotificationBeenSeen = false;
+
+instances.forEach( function ( instanceName ) {
```
You've split editor instances into 3 separate parts. Maybe writing something like:
```js
var editorInstances = CKEDITOR.tools.object.keys( CKEDITOR.instances );
editorInstances.forEach( ... )
```
would be enough?

jacekbogdanski commented on docs/sdk/examples/assets/pastefromoffice/pastefromoffice.js:
```diff
@@ -0,0 +1,164 @@
+let hasNotificationBeenSeen = false;
+
+instances.forEach( function ( instanceName ) {
+    var editor = CKEDITOR.instances[ instanceName ];
+
+    if ( !editor ){
+        return;
+    }
+
+    editor.on( 'instanceReady', function() {
```
There is a small issue with the code. If some instance has been initialized after adding the script to the HTML file, it won't attach events to this specific editor. Instead of looking into `CKEDITOR.instances` and listening on `instance[instanceReady]` event, you could attach the same event to the `CKEDITOR` namespace directly, see https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR.html#event-instanceReady
e.g.
```js
CKEDITOR.on( 'instanceReady', function( evt ) {
  var editor = evt.editor;
  editor.on( 'paste', ... );
} );
```

jacekbogdanski commented on docs/sdk/examples/assets/pastefromoffice/pastefromoffice.js:
```diff
@@ -0,0 +1,164 @@
+
+    editor.on( 'instanceReady', function() {
+        editor.on( 'paste', function( evt ) {
+            var pasteData = CKEDITOR.plugins.pastetools.getClipboardData( evt.data, 'text/html' );
```
I have some small doubts about using `pastetools` here. It makes the dependency on the plugin which should basically not exist in any example which requires this notification. Looking on how `getClipboardData` method works, we can just duplicate it here - https://github.com/ckeditor/ckeditor4/blob/9f0d8ab6463383f89a6c38a40baae792ef15ab5d/plugins/pastetools/plugin.js#L131-L147

jacekbogdanski commented on docs/sdk/examples/assets/pastefromoffice/pastefromoffice.js:
```diff
@@ -0,0 +1,164 @@
+                return;
+            }
+
+            var sourceIndex = getPastedSourceIndex( pasteData );
+            // Show notification only once per pasted source format ( Word, Excel ).
+            if( sourceIndex !== -1 && !hasNotificationBeenSeen && !sources[ sourceIndex ].isAlreadyShow ) {
+                var notice = document.querySelector('.main__notification.notice');
+
+                // Remove notification when user trying to paste content from other source than previous. e.g. Word and Excel.
+                if( notice ){
+                    notice.remove();
+                }
+
+                createClipboardInputNotification( sourceIndex );
+                sources[ sourceIndex ].isAlreadyShow = true;
+            }
```
I have a feeling that the logic responsible for getting a specific handler is a bit complicated. I suppose we could just use for loop for that?

```js
for ( var i = 0; i < sources.length; i++ ) {
  var source = sources[ i ];
  if ( source.regex.test( htmlString ) ) {
    createClipbardInputNotification( source );
    break;
  }
}
```
so we don't have to rely on some magic indexes of `sources` array.

jacekbogdanski commented on docs/sdk/examples/assets/pastefromoffice/pastefromoffice.js:
```diff
@@ -0,0 +1,164 @@
+
+            var sourceIndex = getPastedSourceIndex( pasteData );
+            // Show notification only once per pasted source format ( Word, Excel ).
+            if( sourceIndex !== -1 && !hasNotificationBeenSeen && !sources[ sourceIndex ].isAlreadyShow ) {
```
Not sure, do we need `hasNotificationBeenSeen` if we are updating `isAlreadyShow` property for every format?

jacekbogdanski commented on docs/sdk/examples/assets/pastefromoffice/pastefromoffice.js:
```diff
@@ -0,0 +1,164 @@
+var sources = [
```
Please, add the license information at the top of the file
```
/**
 * @license Copyright (c) 2003-2021, CKSource - Frederico Knabben. All rights reserved.
 * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
 */
```

jacekbogdanski reviewed (changes requested):
It looks like notification is not showing up in some examples:

* `examples/easyimage.html`
* `examples/mentions.html`
* `examples/inline.html`
* `examples/enterkey.html`
* `examples/readonly.html`
* `examples/autotag.html`

Also, integrations may require some special code using [namespaceloaded](http://localhost:9001/ckeditor4/4.16.2/guide/dev_angular.html#namespaceloaded) directive (similar directive is available for every integration)
* `examples/angular.html`
* `examples/react.html`
* `examples/vue.html`
but if you will see that adding this notification to integrations looks complicated, let's sync before working on it as we may not need full coverage for that.

jacekbogdanski commented:
Also, let's make sure that this code also works on older browsers (I suppose docs should open on IE11-9 but it's something we should check). We can add this feature only for modern browsers + IE11, but let's make sure that the script is not breaking examples for older browsers on paste.

KarolDawidziuk commented:
Thanks @jacekbogdanski for the review.

According to your directions I've refactor code and repair the broken examples. 
Unfortunately, the notifications still don't work in the examples with our integrations as this may take longer to handle it.

Changes:
- Refactor js code
- Fixed broken examples
- Fix indentation in HTML files
- Added support to IE9+

jacekbogdanski commented on docs/sdk/examples/assets/pastefromoffice/pastefromoffice.js:
```diff
@@ -0,0 +1,162 @@
+					notice.remove();
+				}
+
+				createClipboardInputNotification( i );
```
Since we already extracted the source, we can pass it directly to the `createClipboardNotification` function.

jacekbogdanski commented on docs/sdk/examples/assets/pastefromoffice/pastefromoffice.js:
```diff
@@ -0,0 +1,162 @@
+		for ( var i = 0; i < sources.length; i++ ) {
+			var source = sources[ i ];
+
+			if( source.regex.test( pasteData ) && !sources[ i ].isAlreadyShow ) {
```
```suggestion
			if( source.regex.test( pasteData ) && !source.isAlreadyShow ) {
```

jacekbogdanski commented on docs/sdk/examples/assets/pastefromoffice/pastefromoffice.js:
```diff
@@ -0,0 +1,162 @@
+				break;
+			}
+		}
+	}, null, null, 9 );
```
I see that setting lower priority worked 🎉 Could you add a small comment on why it's needed? I.e. that `paste` event gets canceled by some plugin custom paste handlers like for `easyimage`?

jacekbogdanski commented on docs/sdk/examples/assets/pastefromoffice/pastefromoffice.js:
```diff
@@ -0,0 +1,162 @@
+/**
```
Note that we are just using this file directly on the page, so every variable or function is available from the global scope if not hidden. E.g. see
![image](https://user-images.githubusercontent.com/10279586/130437712-70492215-91bc-4ea2-9df8-2ac8b0ceede9.png)
To avoid polluting the global namespace (window) we should use IIFE (Immediately Invoked Function Expression) pattern here, e.g.:
```js
( function() {
  // Private variables and functions:
  var sources = [ ... ]
  function x() { ... }
  function y() { ... }
  // Expose public API:
  window.createNotification = function () { ... }
} )();
```

jacekbogdanski reviewed (changes requested):
Looks pretty well, I see that you've fixed the examples mentioned in https://github.com/ckeditor/ckeditor4-docs/pull/366#pullrequestreview-733674212 which is cool. We don't have coverage for integrations but as they are a bit more complicated I think we can skip it for now.

KarolDawidziuk commented:
Thanks, @jacekbogdanski  for the review.

It's ready for the another one.

Changes:
* Avoid polluting global namespace
* Added comment describing add a lower priority than default
* Polishing stuffs
